packages/node/src/dev-server.ts

```diff
@@ -136,77 +147,146 @@ async function serializeRequest(message: IncomingMessage) {
   });
 }
 
-async function createEdgeEventHandler(
-  entrypoint: string
-): Promise<(request: IncomingMessage) => Promise<VercelProxyResponse>> {
-  const buildResult = await ncc(entrypoint, { target: 'es2022' });
-  const userCode = buildResult.code;
-
-  const initialCode = `
-    ${userCode};
-
-    addEventListener('fetch', async (event) => {
-      let serializedRequest = await event.request.text();
-      let requestDetails = JSON.parse(serializedRequest);
+async function compileUserCode(entrypoint: string) {
+  try {
+    const buildResult = await ncc(entrypoint, {
+      target: 'es2022',
+    });
+    const userCode = buildResult.code;
+
+    return `
+      ${userCode};
+
+      addEventListener('fetch', async (event) => {
+        try {
+          let serializedRequest = await event.request.text();
+          let requestDetails = JSON.parse(serializedRequest);
+
+          let body;
+
+          if (requestDetails.method !== 'GET' && requestDetails.method !== 'HEAD') {
+            body = Uint8Array.from(atob(requestDetails.body), c => c.charCodeAt(0));
+          }
+
+          let requestUrl = requestDetails.headers['x-forwarded-proto'] + '://' + requestDetails.headers['x-forwarded-host'] + requestDetails.url;
+
+          let request = new Request(requestUrl, {
+            headers: requestDetails.headers,
+            method: requestDetails.method,
+            body: body
+          });
+
+          event.request = request;
+
+          let edgeHandler = module.exports.default;
+          let response = await edgeHandler(event.request, event);
+
+          return event.respondWith(response);
+        } catch (error) {
+          // we can't easily show a meaningful stack trace
+          // so, stick to just the error message for now
+          event.respondWith(new Response(error.message, {
+            status: 500,
+            headers: {
+              'x-handled-from': 'edge-wrapper'
```

Comment:
This allows us to know that the 500 response comes from our code handling an error in user code instead of the user code intentionally responding with a 500 of their own. There's a test for this if you want more context on what the difference is.